### ISSUE TITLE
Manually modified parameter during trainning, gradients are not being correctly updated. Help

### ISSUE THREAD

**Issue by ithok** [pytorch/pytorch#117967]
## Issue description

Hello.

During training, after replacing a parameter, gradients are not being correctly updated.
Specifically:

There is a fixed tensor adj_mat without gradients.
A mask tensor adj_mask with the same dimensions as adj_mask, initially filled with 0s and 1s, which has gradients.
Element-wise multiplication of adj_mat and adj_mask results in adj.
Forward and backward propagation, along with parameter updates, are performed based on adj (updating adj_mask).
Initially, adj_mask remains constant, but every 20 epochs, I eliminate the smallest 5% of non-zero elements in adj_mask to generate a new adj_mask. This means that every 20 epochs, adj_mask will have fewer non-zero elements by 5%.

However, I observe that after changing adj_mask, during the next batch update, the zeros in adj_mask are also being updated, which is counterintuitive because multiplying any element by zero should not produce any gradient. Can anyone explain why this might be happening and what I should do?

I welcome any assistance from everyone, this is truly crucial for me, and I am deeply grateful.

## Code example

```
def update_norm_adj(self,prune_percent = 0.05):
    # Prune adj_mask, executed every 20 epochs
        with torch.no_grad():
            adj_mask_tensor = self.adj_mask_train.flatten()
            nonzero = torch.abs(adj_mask_tensor) > 0
            adj_mask = adj_mask_tensor[nonzero] 
            adj_total = adj_mask.shape[0]
            adj_y, adj_i = torch.sort(adj_mask.abs())
            adj_thre_index = int(adj_total * prune_percent)
            adj_thre = adj_y[adj_thre_index]
            abs_values = torch.abs(self.adj_mask_train)
            mask = abs_values >= adj_thre
            self.adj_mask_train.data[mask] = 1
            self.adj_mask_train.data[~mask] = 0
        self.pruneflag = True
        print("prune success!")

def forward(self):
        if self.lotteryflag and self.pruneflag
            #In each forward phase, the adjacency matrix is recalculated
            dense_tensor = self.adj_mat
            adj = torch.mul(dense_tensor, self.adj_mask_train)
            self.norm_adj_matrix = self.torch_normalize_adj(adj)

        all_embeddings = self.get_ego_embeddings()  
        feat_mat = self.dropout_sp_mat(self.feat_mat)
        embeddings_list = [all_embeddings]  

        for layer_idx in range(self.n_layers):
            all_embeddings = torch.mm(self.norm_adj_matrix, all_embeddings)
            embeddings_list.append(all_embeddings)
        lightgcn_all_embeddings = torch.stack(embeddings_list, dim=1)
        lightgcn_all_embeddings = torch.mean(lightgcn_all_embeddings, dim=1)

        user_all_embeddings, item_all_embeddings = torch.split(
            lightgcn_all_embeddings, [self.n_users, self.n_items]
        )
        return user_all_embeddings, item_all_embeddings

def torch_normalize_adj(self,adj):
        adj = adj + torch.eye(adj.shape[0]).cuda()
        rowsum = adj.sum(1)
        d_inv_sqrt = torch.pow(rowsum, -0.5).flatten()
        d_inv_sqrt[torch.isinf(d_inv_sqrt)] = 0.0
        d_mat_inv_sqrt = torch.diag(d_inv_sqrt).cuda()
        result = adj.mm(d_mat_inv_sqrt).t().mm(d_mat_inv_sqrt)
        return result

def _train_epoch(self, train_data,aux_train_data, epoch_idx, show_progress=False):
        # Train the model in an epoch

        self.model.train()
        total_loss = None
        iter_data = ( 
            tqdm(
                train_data,
                total=len(train_data),
                ncols=100,
                desc=set_color(f"Train {epoch_idx:>5}", "pink"),
            )
            if show_progress
            else train_data
        )
        if not self.config["single_spec"] and train_data.shuffle:
            train_data.sampler.set_epoch(epoch_idx)

        scaler = amp.GradScaler(enabled=self.enable_scaler)
        batchcount = 0

        dataloader_iterator = iter(aux_train_data)

        for batch_index, interaction in enumerate(iter_data):
            try:
                interaction_aux = next(dataloader_iterator)
            except StopIteration: 
                dataloader_iterator = iter(aux_train_data) #新建一个
                interaction_aux= next(dataloader_iterator)

            interaction = interaction.to(self.device)
            interaction_aux = interaction_aux.to(self.device)
            self.optimizer.zero_grad()
            sync_loss = 0
            if not self.config["single_spec"]:
                self.set_reduce_hook()
                sync_loss = self.sync_grad_loss()
            with torch.autocast(device_type=self.device.type, enabled=self.enable_amp):
                losses = self.model.calculate_loss(interaction, interaction_aux)
            if isinstance(losses, tuple):
                loss = sum(losses)
                loss_tuple = tuple(per_loss.item() for per_loss in losses)
                total_loss = (
                    loss_tuple
                    if total_loss is None
                    else tuple(map(sum, zip(total_loss, loss_tuple)))
                )
            else:
                loss = losses
                total_loss = (
                    losses.item() if total_loss is None else total_loss + losses.item()
                )
            self._check_nan(loss)
            # scaler.scale(loss + sync_loss).backward(retain_graph=True)
            scaler.scale(loss + sync_loss).backward()
            if self.clip_grad_norm:
                clip_grad_norm_(self.model.parameters(), **self.clip_grad_norm)
            scaler.step(self.optimizer)
            scaler.update()
            if self.gpu_available and show_progress:
                iter_data.set_postfix_str(
                    set_color("GPU RAM: " + get_gpu_usage(self.device), "yellow")
                )
            
   

        if self.model.lotteryflag and epoch_idx % 10 == 0 and epoch_idx > 0:
            print("prune!")
            self.model.update_norm_adj()
            self.model.zero_grad()
        return total_loss
```

**Comment by zou3519** [comment]
Do you have a minimal script that reproduces the problem?

**Comment by ithok** [comment]
Thank you!
Due to the complexity of my code, I may not be able to provide directly runnable code, but I'll try to reproduce a minimal script that reproduces the problem:
`|self.user_embedding = torch.nn.Embedding(
     num_embeddings=len(self.user_map), embedding_dim=self.latent_dim
)
self.item_embedding = torch.nn.Embedding(
    num_embeddings=len(self.item_map), embedding_dim=self.latent_dim
)
self.adj_mat = self.get_adj_mat() # 0,1 adjacency matrix, does not contain gradients
self.adj_mask_train = torch.nn.Parameter(self.generate_adj_mask(self.generate_daj_mat()),requires_grad=True)
# 0,1 adjacency matrix mask, contain gradients

optimizer = optim.Adam(params, lr=learning_rate, weight_decay=weight_decay)
optimizer.zero_grad()

user_all_embeddings, item_all_embeddings = self.forward()
u_embeddings = user_all_embeddings[user]
pos_embeddings = item_all_embeddings[pos_item]
neg_embeddings = item_all_embeddings[neg_item]
pos_scores = torch.mul(u_embeddings, pos_embeddings).sum(dim=1)
neg_scores = torch.mul(u_embeddings, neg_embeddings).sum(dim=1)
mf_loss = self.mf_loss(pos_scores, neg_scores)

mf_loss.backward()
optimizer.step()



def forward:
    user_embeddings = self.user_embedding.weight
    item_embeddings = self.item_embedding.weight
    ego_embeddings = torch.cat([user_embeddings, item_embeddings], dim=0)
    embeddings_list = [ego_embeddings]
    if some condition: # prune the adj_mask_train
          with torch.no_grad():
            adj_mask_tensor = self.adj_mask_train.flatten()
            nonzero = torch.abs(adj_mask_tensor) > 0
            adj_mask = adj_mask_tensor[nonzero]  
            adj_total = adj_mask.shape[0]
            adj_y, adj_i = torch.sort(adj_mask.abs())
            adj_thre_index = int(adj_total * prune_percent)
            adj_thre = adj_y[adj_thre_index]

            abs_values = torch.abs(self.adj_mask_train)
            mask = abs_values >= adj_thre
            self.adj_mask_train.data[mask] = 1
            self.adj_mask_train.data[~mask] = 0
    dense_tensor = self.adj_mat
    adj = torch.mul(dense_tensor, self.adj_mask_train)
    self.norm_adj_matrix = self.torch_normalize_adj(adj)
    for layer_idx in range(self.n_layers):
        all_embeddings = torch.mm(self.norm_adj_matrix, all_embeddings)
        embeddings_list.append(all_embeddings)
    lightgcn_all_embeddings = torch.stack(embeddings_list, dim=1)
    lightgcn_all_embeddings = torch.mean(lightgcn_all_embeddings, dim=1)
    user_all_embeddings, item_all_embeddings = torch.split(
                lightgcn_all_embeddings, [self.n_users, self.n_items]
            )
    return user_all_embeddings, item_all_embeddings`
    
 I tried to manually adjust the parameter values, but the gradient calculation was not correct in the backward of the next epoch. The parameters that I reset to zero will still be updated, but this cannot happen because the result obtained by the Hadema product of 0 elements should not have a gradient.

**Comment by ithok** [comment]
> Do you have a minimal script that reproduces the problem?

Hello. I have updated the code。

**Comment by drisspg** [comment]
The script you provided appears to have some issues, without a runnable script it can be quite difficult to debug, would you mind producing something that can be ran via `python script.py`

**Comment by ithok** [comment]
Hello, I wrote similar code and reproduced my problem. Now the code can be run directly.
Due to time reasons, the numerical settings may not be very rigorous, but the problem is the same.
Here the code:
`import torch
import torch.nn as nn
import torch.optim as optim



class SimpleModel(nn.Module):
    def __init__(self):
        super(SimpleModel, self).__init__()
        self.adj = nn.Parameter(torch.Tensor(5, 5).fill_(0.5), requires_grad=False)  # fixed adj
        self.adj_mask = nn.Parameter(torch.ones(5, 5), requires_grad=True)  # trainable adj_mask

    def forward(self):
        masked_adj = torch.mul(self.adj,self.adj_mask) # mask adj with adj_mask
        return masked_adj

def print_tensor_info(tensor):
    num_zeros = (tensor == 0).sum().item()
    num_non_zeros = tensor.numel() - num_zeros
    max_val, min_val = tensor.max(), tensor.min()
    has_nan = torch.isnan(tensor).any().item()

    print(f"Zero values: {num_zeros}")
    print(f"Non-zero values: {num_non_zeros}")
    print(f"Max value: {max_val.item()}")
    print(f"Min value: {min_val.item()}")
    print(f"Has NaN: {has_nan}")
# prune the mask
def prune_mask(mask, percentage):
    # 计算要保留的元素数量
    with torch.no_grad():
        # print("before prune")
        # print_tensor_info(mask)
        adj_mask_tensor = mask.flatten()
        # print("before_adj_mask_tensor.shape: ", adj_mask_tensor.shape)
        nonzero = torch.abs(adj_mask_tensor) > 0
        adj_mask = adj_mask_tensor[nonzero]  # 13264 - 2708
        # print("before_adj_mask.shape: ", adj_mask.shape)
        # print(adj_mask)
        adj_total = adj_mask.shape[0]
        adj_y, adj_i = torch.sort(adj_mask.abs())
        adj_thre_index = int(adj_total * percentage)
        adj_thre = adj_y[adj_thre_index]
        # print("adj_thre", adj_thre)
        abs_values = torch.abs(mask)
        index = abs_values >= adj_thre
        mask.data[index] = 1
        mask.data[~index] = 0
        # print("After prune")
        # print_tensor_info(mask)
    # print("-------")


model = SimpleModel()

optimizer = optim.Adam([model.adj_mask], lr=10)

# input_data = torch.randn(5, 5)
target = torch.randn(5, 5)


pruneflag = False
for epoch in range(41):


    # if pruneflag:
    #     print("Check whether adj_mask has been pruned before forward propagation")
    #     print_tensor_info(model.adj_mask)
    masked_adj = model()
    loss = nn.MSELoss()(masked_adj, target)

    # if pruneflag:
    #     print("Check whether adj_mask has been pruned before backward propagation")
    #     print_tensor_info(model.adj_mask)
    optimizer.zero_grad()
    # if pruneflag:
    #     print("Check whether the adj_mask gradient has been cleared before backpropagation")
    #     print_tensor_info(model.adj_mask.grad)
    loss.backward()
    if(pruneflag):
        print("Check adj_mask gradient after backpropagation")
        print_tensor_info(model.adj_mask.grad)
        # optimizer.zero_grad()
        # print_tensor_info(model.adj_mask.grad)
    optimizer.step()
    if (pruneflag):
        print("Check adj_mask after backpropagation")
        print_tensor_info(model.adj_mask)
        pruneflag = False

    # prune every 20 epoch
    if epoch % 20 == 0:
        print("----epoch" + str(epoch))
        prune_mask(model.adj_mask, 0.2)
        pruneflag = True


`

**Comment by ithok** [comment]
> The script you provided appears to have some issues, without a runnable script it can be quite difficult to debug, would you mind producing something that can be ran via `python script.py`

Hello, I wrote similar code and reproduced my problem. Now the code can be run directly.
Due to time reasons, the numerical settings may not be very rigorous, but the problem is the same.
Here the code:

```
import torch
import torch.nn as nn
import torch.optim as optim



class SimpleModel(nn.Module):
    def __init__(self):
        super(SimpleModel, self).__init__()
        self.adj = nn.Parameter(torch.Tensor(5, 5).fill_(0.5), requires_grad=False)  # fixed adj
        self.adj_mask = nn.Parameter(torch.ones(5, 5), requires_grad=True)  # trainable adj_mask

    def forward(self):
        masked_adj = torch.mul(self.adj,self.adj_mask) # mask adj with adj_mask
        return masked_adj

def print_tensor_info(tensor):
    num_zeros = (tensor == 0).sum().item()
    num_non_zeros = tensor.numel() - num_zeros
    max_val, min_val = tensor.max(), tensor.min()
    has_nan = torch.isnan(tensor).any().item()

    print(f"Zero values: {num_zeros}")
    print(f"Non-zero values: {num_non_zeros}")
    print(f"Max value: {max_val.item()}")
    print(f"Min value: {min_val.item()}")
    print(f"Has NaN: {has_nan}")

# prune the mask
def prune_mask(mask, percentage):
    with torch.no_grad():
        # print("before prune")
        # print_tensor_info(mask)
        adj_mask_tensor = mask.flatten()
        # print("before_adj_mask_tensor.shape: ", adj_mask_tensor.shape)
        nonzero = torch.abs(adj_mask_tensor) > 0
        adj_mask = adj_mask_tensor[nonzero]  # 13264 - 2708
        # print("before_adj_mask.shape: ", adj_mask.shape)
        # print(adj_mask)
        adj_total = adj_mask.shape[0]
        adj_y, adj_i = torch.sort(adj_mask.abs())
        adj_thre_index = int(adj_total * percentage)
        adj_thre = adj_y[adj_thre_index]
        # print("adj_thre", adj_thre)
        abs_values = torch.abs(mask)
        index = abs_values >= adj_thre
        mask.data[index] = 1
        mask.data[~index] = 0
        # print("After prune")
        # print_tensor_info(mask)
    # print("-------")


model = SimpleModel()

optimizer = optim.Adam([model.adj_mask], lr=10)

# input_data = torch.randn(5, 5)
target = torch.randn(5, 5)


pruneflag = False
for epoch in range(41):


    # if pruneflag:
    #     print("Check whether adj_mask has been pruned before forward propagation")
    #     print_tensor_info(model.adj_mask)
    masked_adj = model()
    loss = nn.MSELoss()(masked_adj, target)

    # if pruneflag:
    #     print("Check whether adj_mask has been pruned before backward propagation")
    #     print_tensor_info(model.adj_mask)
    optimizer.zero_grad()
    # if pruneflag:
    #     print("Check whether the adj_mask gradient has been cleared before backpropagation")
    #     print_tensor_info(model.adj_mask.grad)
    loss.backward()
    if(pruneflag):
        print("Check adj_mask gradient after backpropagation")
        print_tensor_info(model.adj_mask.grad)
        # optimizer.zero_grad()
        # print_tensor_info(model.adj_mask.grad)
    optimizer.step()
    if (pruneflag):
        print("Check adj_mask after backpropagation")
        print_tensor_info(model.adj_mask)
        pruneflag = False

    # prune every 20 epoch
    if epoch % 20 == 0:
        print("----epoch" + str(epoch))
        prune_mask(model.adj_mask, 0.2)
        pruneflag = True



```


After modifying the adj_mask tensor, during the backpropagation, I observe that the gradients on the adj_mask are not updating as intended—specifically, elements multiplied by zero through a Hadamard product are unexpectedly receiving gradients, which results in incorrect gradient computations.

Furthermore, despite manually setting the gradients to zero, the adj_mask continues to get updated following optimizer.step() in my code (you can refer to the commented out parts of my code). I hypothesize that this could be related to the momentum inherent in the Adam optimizer because the issue does not occur when I switch to SGD. Nonetheless, this hypothesis doesn't clarify the underlying reason for the first question.

Thank you in advance for taking the time to read through my query and for any assistance you can provide.
